src/inference/src/dev/make_tensor.cpp

```diff
@@ -178,28 +180,78 @@ class AllocatedTensor : public ViewTensor {
                      shape,
                      [&] {
                          OPENVINO_ASSERT(allocator, "Allocator was not initialized");
-                         return const_cast<Allocator&>(allocator).allocate(element_type.size() * shape_size(shape));
+                         auto num_elements = shape_size(shape);
+                         auto data = const_cast<Allocator&>(allocator).allocate(element_type.size() * num_elements);
+                         init(data, element_type, shape);
+                         return data;
                      }()},
           m_allocator{allocator} {}
 
     ~AllocatedTensor() {
+        destroy();
         m_allocator.deallocate(m_ptr, get_byte_size());
     }
 
     void set_shape(ov::Shape new_shape) override {
         if (m_shape == new_shape)
             return;
+
         auto old_byte_size = get_byte_size();
+        auto old_shape = m_shape;
         m_shape = std::move(new_shape);
+
+        if (m_element_type == element::string) {
+            auto new_num_elements = shape_size(new_shape);
+            auto num_elements = shape_size(old_shape);
+            auto data = static_cast<std::string*>(m_ptr);
+            if (get_byte_size() > old_byte_size) {
+                // expect memory re-allocation that is why destruct all std::string objects
+                for (size_t ind = 0; ind < num_elements; ++ind) {
+                    using std::string;
+                    data[ind].~string();
+                }
+            } else {
+                // re-use already allocated memory for ov::Tensor of new shape
+                OPENVINO_ASSERT(
+                    new_num_elements <= num_elements,
+                    "New number of elements is expected to be less than the current number in string ov::Tensor");
+                // remaining unneeded objects need to be destructed
+                for (size_t ind = new_num_elements; ind < num_elements; ++ind) {
+                    using std::string;
+                    data[ind].~string();
+                }
+            }
```

Comment:
Consider to extract it as private helper which do dedicated re-allocation for `string`?
Set shape could look cleaner.

Reply:
https://github.com/openvinotoolkit/openvino/pull/21244#discussion_r1408980470